PR for review

grimmier378 opened the pull request.

grimmier378 committed:
merged all commits into a single commit on dev branch for review

brainiac commented on Theme.h:
```diff
@@ -0,0 +1,488 @@
+* @brief Pop a value from the ImGui context
+* @param int popCounts The number of colors pushed + (the number of styles pushed << 16)
+*/
+void PopTheme(int popCounts)
+{
+	int colorCount = popCounts & 0xFFFF;
+	int styleCount = popCounts >> 16;
+	ImGui::PopStyleColor(colorCount);
+	ImGui::PopStyleVar(styleCount);
+}
```
just pass two ints, or a struct with two uint16s, this usage is not intuitive and buys you nothing.

brainiac reviewed (commented):
Reviewed the Theme.h

If you make all the functions there inline, and your themeSet static inline, then you'll be better off as something to be used as header-only.

brainiac commented on MQ2GrimGUI.cpp:
```diff
@@ -1,505 +1,1510 @@
+				s_NumSettings.secondAggroPct = pAggroInfo->aggroData[AD_Secondary].AggroPct;
+				
+				if (pAggroInfo->AggroSecondaryID)
+					s_SecondAggroName = GetSpawnByID(pAggroInfo->AggroSecondaryID)->DisplayedName;
```
Definitely want to check that this didn't return null.

brainiac commented on MQ2GrimGUI.cpp:
```diff
@@ -1,505 +1,1510 @@
+
 			g_LastUpdateTime = now;
+
+			s_DanNetEnabled = mq::IsPluginLoaded("MQ2DanNet");
```
use this variable in `DrawGroupWindow`. If you don't want to check _every frame_ that MQ2DanNet is installed, you can listen for it being loaded/unloaded in the OnLoadPlugin/OnUnloadPlugin callbacks.

brainiac commented on MQ2GrimGUI.cpp:
```diff
@@ -1,505 +1,1510 @@
-			b_flashCombat = !b_flashCombat;
-			g_LastFlashTime = now;
+			s_MemSpellName = pSpellPicker->SelectedSpell->Name;
+			MemSpell(pLocalPlayer, fmt::format("{} \"{}\"", s_MemGemIndex, s_MemSpellName).c_str());
```
This should be `DoCommandf("/memspell %s \"%s\"", s_MemGemIndex, s_MemSpellName)`

MemSpell is a command function. You shouldn't even be able to call it from here ...

brainiac commented on MQ2GrimGUI.cpp:
```diff
@@ -1,505 +1,1510 @@
+	pSpellInspector = new grimgui::SpellsInspector();
+	pSpellPicker = new SpellPicker();
+
+	static bool s_DefaultExists = std::filesystem::exists(DEFAULT_INI);
```
static is unnecessary. Function is only called once. Need the std::error_code too.

brainiac commented on MQ2GrimGUI.cpp:
```diff
@@ -1,505 +1,1510 @@
+	DebugSpewAlways("Shutting down MQ2GrimGUI");
+	RemoveCommand("/grimui");
+	pSpellPicker = nullptr;
+	pSpellInspector = nullptr;
```
you need to `delete` these if you `new` them.

brainiac reviewed (changes requested):
Lots of comments left. Here is a general summary

- [ ] Avoid unnecessary `std::string` creations. mq has a ton of efficient string helper functions in mq/base/String.h that work on string_view so they work on any string-like variable. Lean heavily into these.
- [ ] Look for all the places that you did `&something[0]` and fix not take address of first element. This should never be necessary, or else you're doing something unsafe.
- [ ] Lots of places where you're getting a spawn pointer from one variable and sticking into another temporary variable that is never written. Don't do that. `if (PSPAWNINFO pCharInfo = pLocalPlayer)` just use `pLocalPlayer` directly.
- [ ] Don't use `PSPAWNINFO` it is `PlayerClient*`. Don't use `GetCharInfo()`, use `pLocalPC`.
- [ ] std::filesystem calls need std::error_code or else you need exception handling.
- [ ] ImGui IDs need a look over. You generally only need to provide them for your top level windows and then only elements that share the same name in the same scope. PushID makes everything underneath unique for things like table rows or for loops. Adding ## everywhere when it isn't needed just adds more than is necessary.
- [ ] Whenever you `new` a pointer you're on the hook to `delete` it. Otherwise you have memory leaks.
- [ ] Your usage of OnLoadPlugin / OnUnloadPlugin is critically wrong and must be addressed.
- [ ] There is way too much in `MQ2GrimGUI.h`. It really isn't even a header file. Just take everything and dump it into the .cpp file. At some point when you want to split into multiple .cpp files, only add things they both need in common as forward declarations to a new header file.
- [ ] `ImGui::MenuItem` will flip bools for you, to make your code simpler.
- [ ] `mq::imgui::SpellGem` may make your life easier. As will `mq::imgui::HelpMarker`.
- [ ] Don't call `pSpawnManager->GetSpawnByID`. Use the MQ version.
- [ ] `EzCommand(fmt::format(...))` -> `DoCommandf`
- [ ] Talk to me about the spell inspector if there is something you want to re-use. Copy pasting parts of it is not okay with me if it can be reused.
- [ ] Theme code should just be replaced with ImGuiStyle. I left comments on your theme but I think a lot of this could potentially just go away.
- [ ] As we discussed in discord, SpellPicker shouldn't be copying spell data. It makes this much more complex for you.
- [ ] Clean up your header inclusion order to help keep your dependencies clean.

brainiac commented:
Oh also, I might encourage you to remove the "2" from the name of this plugin!

grimmier378 committed:
refactored theme.h

grimmier378 committed:
cleaned up spell picker

kept the table so we can sort the tree view and group by category/subcategory  sorted by lvl highest to lowest.

Moved the sorting and filtering out of the imgui frames.

grimmier378 committed:
name change on plugin dropped the 2

grimmier378 committed:
removed header file and cleaned up main cpp

I think I got all of the suggestions. I may have missed something still.

Removed a lot of string compares and strings for char* and use string_view where appropriate.

grimmier378 committed:
theme include cleanup

grimmier378 committed:
whitespace and alignment

brainiac commented on Theme.h:
```diff
@@ -0,0 +1,470 @@
+
+inline void StyleColorsGrape()
+	{
+	ImGuiStyle*                            style  = &ImGui::GetStyle();
+	ImVec4*                                colors = style-> Colors;
```
This is replacing the existing style. You don't want to do that. You want to create your own ImGuiStyle.

Applying your style simply becomes replacing the current style with your own while you are inside your window code, and then restoring the old style.

something like...

```
ImGuiStyle oldStyle = ImGui::GetStyle();
ImGuui::GetStyle() = StyleColorsGrape();

... window code

ImGui::GetSTyle() = oldStyle
```

I think that shuld work

grimmier378 replied:
ahh ok will try that when i get home. 
I found the ones from style editor drop down and thought that was the way to do it.

grimmier378 committed:
theme update.

store originalTheme and replace instead of overwriting every time.

grimmier378 committed:
cleaned up default

just return the original theme if default no need for an extra function.

grimmier378 committed:
fix broken default bar colors.

grimmier378 committed:
fix inverted check when camping to switch config

grimmier378 committed:
tweaked night theme

grimmier378 committed:
fixed m_ names

grimmier378 committed:
prevent windows from stealing focus on appearing

grimmier378 committed:
window position saving.

grimmier378 committed:
fix spell inspect

grimmier378 committed:
save docking state so we don't apply position and size

grimmier378 committed:
locking menu fix

grimmier378 committed:
menu tweaks

grimmier378 committed:
lock icons are actual buttons.

grimmier378 committed:
some cleanup

grimmier378 committed:
WIP on dev2

grimmier378 committed:
cotton candy theme

grimmier378 committed:
theme work

all lookups based on index removed the string save and lookup compares.
optional theme rounding for each theme.
checkboxes for settings.

grimmier378 committed:
new themes from button master

reworked lime

grimmier378 committed:
casting time fix

was subtracting a negative.

grimmier378 committed:
fix follow/come buttons and imgui menu
